### ISSUE TITLE
[Request] Add RTSPT scheme

### ISSUE THREAD

**Issue by Fummowo** [androidx/media#1484]
If you know that an RTSP stream is TCP only, it would be nice to have a user-friendly way to specify that protocol, rather than having to rely on and wait for the fall back from UDP. Many ISP's block RTSP UDP traffic, so even when https://github.com/androidx/media/issues/586 is resolved, there would still be a noticeable delay before the stream starts for many people.

### Proposed solution
Add a dedicated RTSPT scheme to force use of the RTSP TCP protocol. (rtspt://live.stream)
This would also allow the user to expect a more consistent result across platforms (e.g with AVPro-Video) as this is a solution also used in Media Foundation on Windows.

**Comment by Arcitec** [comment]
@microkatz This is a major improvement if implemented.

The RTSP protocol handlers are:

- rtsp:// for auto detection.
- rtspu:// for forcing UDP only.
- rtspt:// for forcing TCP only.

The most common way to specify RTSP streams is rtspt://, probably because TCP is the only RTSP transport method that works over the internet.

**Comment by microkatz** [comment]
@Fummowo 

Apologies for the delay. If one knows that only TCP is supported, one can configure the RtspMediaSource.Factory to only use TCP. https://github.com/androidx/media/blob/release/libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/RtspMediaSource.java#L95

Hope that helps you in the short term.

@Arcitec 

Thank you for signaling to this issue! It does help bring tickets back onto our radar. As far as the request, I do see from your posted link that a similar request was made in the other project. However, is this an industry standard for the uri prefix? We are more likely to implement if there is an RFC spec for RTSPT that is supporting this.

**Comment by Arcitec** [comment]
@microkatz Thank you again for the advice to set the `setForceUseRtpTcp` boolean to force TCP mode. It's definitely a great workaround.

Regarding the RFCs, I looked at both of them linked here and don't see any formal specification for the protocol handlers:

https://en.wikipedia.org/wiki/Real-Time_Streaming_Protocol

I am not sure where the `rtpst://` and `rtspu://` originated, but I know that they are widely supported.

Here's [Microsoft's table of Media Foundation player protocols](https://learn.microsoft.com/en-us/windows/win32/medfound/supported-protocols#protocol-rollover):

![image](https://github.com/user-attachments/assets/1679e97f-0865-46ab-9f79-217abd5c8e2b)

The practice of using `rtspt://` may have originated with the GStreamer project. It also mentions how that's implemented by sending a `SETUP` request that requests TCP mode, as seen in this comment:

https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/2412#note_1839999

So it might not be an official part of the RTSP spec, and more of a streaming industry "hey that was a good idea, let's add a way for the URLs to force TCP mode".

What are your thoughts on this so far?

**Comment by microkatz** [comment]
@Arcitec @Fummowo 

Thank you for the detailed response. I started looking into implementing the feature. I ran into an issue where the RTSP servers we have do not accept requests with `rtspt//`. Would you have a publicly available server that we can use to test the feature?

**Comment by Fummowo** [comment]
@microkatz Apologies for the late response,
I believe any RTSP server using TCP should accept it, as the request is still sent using `rtsp://`, the RTSPT scheme is only supposed to force the client to connect using TCP. However, a publicly available server that works with other implementations of this scheme is available at `rtspt://stream.vrcdn.live/live/vrcdn`.

**Comment by microkatz** [comment]
@Fummowo 
If you were able to setup a server using that service and then email the necessary information for us to stream to android-media-github@google.com with the subject `Issue #1484` then that would be helpful!

Are you saying that even though the requested URI to stream from has the scheme "rtspt", ExoPlayer should alter the provided URI to "rtsp" for use in all interactions? In your testing, does the server fail if the rtsp messages are sent with the "rtspt" scheme. This ambiguity is why we usually try to adhere to an official spec if possible.

**Comment by Fummowo** [comment]
@microkatz 
Yes, the URI should be altered since servers won't be made to accept `rtspt` because it is not part of the official spec, as seen from testing.
This scheme is used to force the client to connect with TCP by slightly altering the URI.

**Comment by microkatz** [comment]
@Fummowo 
The RTSPT feature has been implemented in the main branch, https://github.com/androidx/media/commit/a220b0cb5e16e6cb8d8719a0e1ffa7e00859abc0. Hopefully this works for you. Re-open or create a new ticket if there are any issues.
